File: examples/go/shrmpl/shrmpl.go
package shrmpl

import (
	"bufio"
	"crypto/tls"
	"fmt"
	"io"
	"net"
	"net/http"
	"strconv"
	"strings"
	"time"
)

// KV Client
type KVClient struct {
	host string
	port int
	conn net.Conn
}

// KVListItem represents a key-value pair with optional expiration
type KVListItem struct {
	Key       string
	Value     string
	ExpiresAt *int64 // Unix timestamp, nil if no expiration
}

func NewKVClient(host string, port int) *KVClient {
	return &KVClient{
		host: host,
		port: port,
	}
}

func (c *KVClient) Connect() (bool, error) {
	addr := fmt.Sprintf("%s:%d", c.host, c.port)
	conn, err := net.DialTimeout("tcp", addr, 5*time.Second)
	if err != nil {
		return false, err
	}
	
	// Set TCP_NODELAY
	if tcpConn, ok := conn.(*net.TCPConn); ok {
		tcpConn.SetNoDelay(true)
	}
	
	c.conn = conn
	return true, nil
}

func (c *KVClient) sendCommand(cmd string) (string, error) {
	if c.conn == nil {
		return "", fmt.Errorf("not connected")
	}
	
	// Send command
	_, err := c.conn.Write([]byte(cmd + "\n"))
	if err != nil {
		return "", err
	}
	
	// Read response
	reader := bufio.NewReader(c.conn)
	response, err := reader.ReadString('\n')
	if err != nil {
		return "", err
	}
	
	response = strings.TrimSpace(response)
	
	// Filter out heartbeats
	if response == "UPONG" {
		return "", fmt.Errorf("heartbeat received")
	}
	if response == "TERM" {
		return "", fmt.Errorf("server shutting down")
	}
	
	return response, nil
}

func (c *KVClient) Get(key string) (string, error) {
	if len(key) > 100 {
		return "", fmt.Errorf("key length exceeds 100 characters")
	}
	
	response, err := c.sendCommand(fmt.Sprintf("GET %s", key))
	if err != nil {
		if strings.Contains(err.Error(), "key not found") {
			return "", nil // Key not found is not an error
		}
		return "", err
	}
	
	if strings.HasPrefix(response, "ERROR") {
		return "", fmt.Errorf(response)
	}
	
	return response, nil
}

func (c *KVClient) Set(key, value string, ttl string) (bool, error) {
	if len(key) > 100 || len(value) > 100 {
		return false, fmt.Errorf("key or value length exceeds 100 characters")
	}
	
	var cmd string
	if ttl != "" {
		cmd = fmt.Sprintf("SET %s %s %s", key, value, ttl)
	} else {
		cmd = fmt.Sprintf("SET %s %s", key, value)
	}
	
	response, err := c.sendCommand(cmd)
	if err != nil {
		return false, err
	}
	
	return response == "OK", nil
}

func (c *KVClient) Incr(key string, ttl string) (int, error) {
	if len(key) > 100 {
		return 0, fmt.Errorf("key length exceeds 100 characters")
	}
	
	var cmd string
	if ttl != "" {
		cmd = fmt.Sprintf("INCR %s %s", key, ttl)
	} else {
		cmd = fmt.Sprintf("INCR %s", key)
	}
	
	response, err := c.sendCommand(cmd)
	if err != nil {
		return 0, err
	}
	
	if strings.HasPrefix(response, "ERROR") {
		return 0, fmt.Errorf(response)
	}
	
	result, err := strconv.Atoi(response)
	if err != nil {
		return 0, fmt.Errorf("invalid response: %s", response)
	}
	
	return result, nil
}

func (c *KVClient) Delete(key string) (bool, error) {
	if len(key) > 100 {
		return false, fmt.Errorf("key length exceeds 100 characters")
	}
	
	response, err := c.sendCommand(fmt.Sprintf("DEL %s", key))
	if err != nil {
		if strings.Contains(err.Error(), "key not found") {
			return false, nil // Key not found is not an error
		}
		return false, err
	}
	
	return response == "OK", nil
}

func (c *KVClient) Ping() (bool, error) {
	response, err := c.sendCommand("PING")
	if err != nil {
		return false, err
	}
	
	return response == "PONG", nil
}

func (c *KVClient) List() ([]KVListItem, error) {
	response, err := c.sendCommand("LIST")
	if err != nil {
		return nil, err
	}
	
	if strings.HasPrefix(response, "ERROR") {
		return nil, fmt.Errorf(response)
	}
	
	var items []KVListItem
	if strings.TrimSpace(response) == "" {
		return items, nil
	}
	
	lines := strings.Split(response, "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		
		// Parse format: key=value,expiration
		parts := strings.SplitN(line, "=", 3)
		if len(parts) != 3 {
			continue
		}
		
		key := parts[0]
		valueAndExpiration := parts[2]
		
		// Split value and expiration
		valueParts := strings.SplitN(valueAndExpiration, ",", 2)
		if len(valueParts) != 2 {
			continue
		}
		
		value := valueParts[0]
		expirationStr := valueParts[1]
		
		var expiresAt *int64
		if expirationStr != "no-expiration" {
			if timestamp, err := strconv.ParseInt(expirationStr, 10, 64); err == nil {
				expiresAt = &timestamp
			}
		}
		
		items = append(items, KVListItem{
			Key:       key,
			Value:     value,
			ExpiresAt: expiresAt,
		})
	}
	
	return items, nil
}

func (c *KVClient) Close() {
	if c.conn != nil {
		c.conn.Close()
		c.conn = nil
	}
}

// Log Client
type LogClient struct {
	host string
	port int
	conn net.Conn
}

func NewLogClient(host string, port int) *LogClient {
	return &LogClient{
		host: host,
		port: port,
	}
}

func (c *LogClient) Connect() (bool, error) {
	addr := fmt.Sprintf("%s:%d", c.host, c.port)
	conn, err := net.DialTimeout("tcp", addr, 5*time.Second)
	if err != nil {
		return false, err
	}
	
	// Set TCP_NODELAY
	if tcpConn, ok := conn.(*net.TCPConn); ok {
		tcpConn.SetNoDelay(true)
	}
	
	c.conn = conn
	return true, nil
}

func (c *LogClient) Send(level, host, code, message string) error {
	// Validate inputs
	if len(level) != 4 {
		return fmt.Errorf("level must be exactly 4 characters")
	}
	if len(host) > 32 {
		return fmt.Errorf("host must be <= 32 characters")
	}
	if len(code) != 4 {
		return fmt.Errorf("code must be exactly 4 characters")
	}
	if len(message) > 4096 {
		return fmt.Errorf("message must be <= 4096 characters")
	}
	
	// Format: [LVL(4)] [HOST(32)] [CODE(4)] [LEN(4)]: [MSG]\n
	paddedHost := fmt.Sprintf("%-32s", host[:32])
	paddedLevel := fmt.Sprintf("%-4s", level[:4])
	paddedCode := fmt.Sprintf("%-4s", code[:4])
	msgLen := fmt.Sprintf("%04d", len(message))
	
	logLine := fmt.Sprintf("[%s] [%s] [%s] [%s]: %s\n", paddedLevel, paddedHost, paddedCode, msgLen, message)
	
	_, err := c.conn.Write([]byte(logLine))
	return err
}

func (c *LogClient) Close() {
	if c.conn != nil {
		c.conn.Close()
		c.conn = nil
	}
}

// Vault Client
type VaultClient struct {
	serverURL string
	certPath  string
	keyPath   string
	secret    string
	client    *http.Client
}

func NewVaultClient(serverURL, certPath, keyPath, secret string) *VaultClient {
	return &VaultClient{
		serverURL: strings.TrimRight(serverURL, "/"),
		certPath:  certPath,
		keyPath:   keyPath,
		secret:    secret,
	}
}

func (c *VaultClient) Connect() (bool, error) {
	// Load client certificates
	cert, err := tls.LoadX509KeyPair(c.certPath, c.keyPath)
	if err != nil {
		return false, fmt.Errorf("failed to load certificates: %v", err)
	}
	
	// Create TLS config
	tlsConfig := &tls.Config{
		Certificates: []tls.Certificate{cert},
	}
	
	// Create HTTP client
	transport := &http.Transport{
		TLSClientConfig: tlsConfig,
	}
	c.client = &http.Client{
		Transport: transport,
		Timeout:   10 * time.Second,
	}
	
	// Connection setup successful - actual testing happens during GetConfig calls
	return true, nil
}

func (c *VaultClient) GetConfig(filename string) (string, error) {
	if c.client == nil {
		return "", fmt.Errorf("not connected")
	}
	
	url := fmt.Sprintf("%s/%s?secret=%s", c.serverURL, filename, c.secret)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return "", err
	}
	
	resp, err := c.client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	
	switch resp.StatusCode {
	case 200:
		content, err := io.ReadAll(resp.Body)
		return string(content), err
	case 404:
		return "", fmt.Errorf("file not found")
	case 401:
		return "", fmt.Errorf("unauthorized - invalid certificate or secret")
	case 429:
		return "", fmt.Errorf("rate limit exceeded")
	default:
		return "", fmt.Errorf("HTTP error: %d", resp.StatusCode)
	}
}

File: examples/python/shrmpl.py
#!/usr/bin/env python3
"""
Shrmpl Client Library

Provides client classes for KV, Log, and Vault services.
Each service maintains persistent connections and returns (result, error) tuples.
"""

import socket
import ssl
import urllib.request
import urllib.error
from typing import Optional, Tuple, Union


class ShrmplKV:
    """Client for Shrmpl KV Server"""
    
    def __init__(self, host: str, port: int):
        self.host = host
        self.port = port
        self.socket = None
        self.reader = None
        self.writer = None
    
    def connect(self) -> Tuple[bool, Optional[str]]:
        """Connect to KV server"""
        try:
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket.settimeout(5)
            self.socket.connect((self.host, self.port))
            self.socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
            return True, None
        except Exception as e:
            return False, str(e)
    
    def _send_command(self, command: str) -> Tuple[Optional[str], Optional[str]]:
        """Send command and get response"""
        if not self.socket:
            return None, "Not connected"
        
        try:
            # Send command
            self.socket.sendall(f"{command}\n".encode())
            
            # Read response
            response = ""
            while True:
                chunk = self.socket.recv(1024).decode()
                if not chunk:
                    break
                response += chunk
                if response.endswith('\n'):
                    break
            
            response = response.strip()
            
            # Filter out heartbeats
            if response == "UPONG":
                return None, "Heartbeat received"
            elif response == "TERM":
                return None, "Server shutting down"
            elif response.startswith("ERROR"):
                return None, response
            else:
                return response, None
                
        except Exception as e:
            return None, str(e)
    
    def get(self, key: str) -> Tuple[Optional[str], Optional[str]]:
        """Get value for key"""
        if len(key) > 100:
            return None, "Key length exceeds 100 characters"
        
        result, error = self._send_command(f"GET {key}")
        if error:
            if "key not found" in str(error):
                return None, None  # Key not found is not an error
            return None, error
        return result, None
    
    def set(self, key: str, value: str, ttl: Optional[str] = None) -> Tuple[bool, Optional[str]]:
        """Set key to value with optional TTL"""
        if len(key) > 100 or len(value) > 100:
            return False, "Key or value length exceeds 100 characters"
        
        if ttl:
            command = f"SET {key} {value} {ttl}"
        else:
            command = f"SET {key} {value}"
        
        result, error = self._send_command(command)
        if error:
            return False, error
        return result == "OK", None
    
    def incr(self, key: str, ttl: Optional[str] = None) -> Tuple[Optional[int], Optional[str]]:
        """Increment integer value by 1 with optional TTL"""
        if len(key) > 100:
            return None, "Key length exceeds 100 characters"
        
        if ttl:
            command = f"INCR {key} {ttl}"
        else:
            command = f"INCR {key}"
        
        result, error = self._send_command(command)
        if error:
            return None, error
        
        if result is None:
            return None, "No response received"
        
        try:
            return int(result), None
        except (ValueError, TypeError):
            return None, f"Invalid response: {result}"
    
    def delete(self, key: str) -> Tuple[bool, Optional[str]]:
        """Delete key"""
        if len(key) > 100:
            return False, "Key length exceeds 100 characters"
        
        result, error = self._send_command(f"DEL {key}")
        if error:
            if "key not found" in str(error):
                return False, None  # Key not found is not an error
            return False, error
        return result == "OK", None
    
    def ping(self) -> Tuple[bool, Optional[str]]:
        """Ping server"""
        result, error = self._send_command("PING")
        if error:
            return False, error
        return result == "PONG", None
    
    def list(self) -> Tuple[Optional[list], Optional[str]]:
        """List all keys with values and expiration timestamps"""
        result, error = self._send_command("LIST")
        if error:
            return None, error
        
        if result is None:
            return [], None
        
        items = []
        if result.strip() == "":
            return [], None
        
        for line in result.strip().split('\n'):
            if line.strip() == "":
                continue
            
            parts = line.split('=', 2)
            if len(parts) != 3:
                continue
            
            key = parts[0]
            value_and_expiration = parts[2]
            value_parts = value_and_expiration.split(',', 1)
            
            if len(value_parts) != 2:
                continue
            
            value = value_parts[0]
            expiration_str = value_parts[1]
            
            if expiration_str == "no-expiration":
                expiration = None
            else:
                try:
                    expiration = int(expiration_str)
                except ValueError:
                    expiration = None
            
            items.append((key, value, expiration))
        
        return items, None
    
    def close(self):
        """Close connection"""
        if self.socket:
            self.socket.close()
            self.socket = None


class ShrmplLog:
    """Client for Shrmpl Log Server"""
    
    def __init__(self, host: str, port: int):
        self.host = host
        self.port = port
        self.socket = None
    
    def connect(self) -> Tuple[bool, Optional[str]]:
        """Connect to Log server"""
        try:
            self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.socket.settimeout(5)
            self.socket.connect((self.host, self.port))
            self.socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
            return True, None
        except Exception as e:
            return False, str(e)
    
    def send(self, level: str, host: str, code: str, message: str) -> Tuple[bool, Optional[str]]:
        """Send log message"""
        try:
            # Validate inputs
            if len(level) != 4:
                return False, "Level must be exactly 4 characters"
            if len(host) > 32:
                return False, "Host must be <= 32 characters"
            if len(code) != 4:
                return False, "Code must be exactly 4 characters"
            if len(message) > 4096:
                return False, "Message must be <= 4096 characters"
            
            # Format: [LVL(4)] [HOST(32)] [CODE(4)] [LEN(4)]: [MSG]\n
            padded_host = host.ljust(32)[:32]  # Pad or truncate to 32 chars
            padded_level = level.ljust(4)[:4]    # Pad or truncate to 4 chars
            padded_code = code.ljust(4)[:4]      # Pad or truncate to 4 chars
            msg_len = f"{len(message):04d}"     # Zero-padded length
            
            log_line = f"[{padded_level}] [{padded_host}] [{padded_code}] [{msg_len}]: {message}\n"
            
            if self.socket:
                self.socket.sendall(log_line.encode())
            return True, None
            
        except Exception as e:
            return False, str(e)
    
    def close(self):
        """Close connection"""
        if self.socket:
            self.socket.close()
            self.socket = None


class ShrmplVault:
    """Client for Shrmpl Vault Server"""
    
    def __init__(self, server_url: str, cert_path: str, key_path: str, secret: str):
        self.server_url = server_url.rstrip('/')
        self.cert_path = cert_path
        self.key_path = key_path
        self.secret = secret
    
    def connect(self) -> Tuple[bool, Optional[str]]:
        """Validate certificates and test connection"""
        try:
            # Create SSL context
            context = ssl.create_default_context()
            context.check_hostname = False
            context.verify_mode = ssl.CERT_REQUIRED
            
            # Load client certificates
            context.load_cert_chain(self.cert_path, self.key_path)
            
            # Test connection with a simple request
            test_url = f"{self.server_url}/test?secret={self.secret}"
            req = urllib.request.Request(test_url)
            
            try:
                with urllib.request.urlopen(req, context=context, timeout=10) as response:
                    if response.status == 404:
                        # 404 is expected for test file, means connection works
                        return True, None
                    else:
                        return False, f"Test failed with status: {response.status}"
            except urllib.error.HTTPError as e:
                if e.code == 404:
                    # 404 is expected for test file
                    return True, None
                else:
                    return False, f"HTTP error: {e.code} - {e.reason}"
                    
        except Exception as e:
            return False, str(e)
    
    def get_config(self, filename: str) -> Tuple[Optional[str], Optional[str]]:
        """Get configuration file from vault"""
        try:
            # Create SSL context
            context = ssl.create_default_context()
            context.check_hostname = False
            context.verify_mode = ssl.CERT_REQUIRED
            context.load_cert_chain(self.cert_path, self.key_path)
            
            # Make request
            url = f"{self.server_url}/{filename}?secret={self.secret}"
            req = urllib.request.Request(url)
            
            with urllib.request.urlopen(req, context=context, timeout=10) as response:
                if response.status == 200:
                    content = response.read().decode()
                    return content, None
                elif response.status == 404:
                    return None, "File not found"
                elif response.status == 401:
                    return None, "Unauthorized - invalid certificate or secret"
                elif response.status == 429:
                    return None, "Rate limit exceeded"
                else:
                    return None, f"HTTP error: {response.status}"
                    
        except urllib.error.HTTPError as e:
            if e.code == 404:
                return None, "File not found"
            elif e.code == 401:
                return None, "Unauthorized - invalid certificate or secret"
            elif e.code == 429:
                return None, "Rate limit exceeded"
            else:
                return None, f"HTTP error: {e.code} - {e.reason}"
        except Exception as e:
            return None, str(e)